feat(rhineng-12924): Proposed mechanism for replicated table migrations

chambridge opened the pull request:
# Overview

This PR is being created to address [RHINENG-12924](https://issues.redhat.com/browse/RHINENG-12924).
* Define migration runner script that would be utilized by the application team
* Create app migrations folder that will house the SQL files for a migration
* Capture migration strategy in a README document in the new folder

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Include raw query examples in the PR description, if adding/modifying SQL query
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

chambridge commented:
/retest

thearifismail commented:
@chambridge Reference to this working should be provided in the main/first README.  Should this wait until logical replication has been enabled by the host-inventory first?

chambridge commented:
@fstavela I've updated the main README to link to this documentation. I've also added a make command for generating the SQL script.

I like what you are proposing for something more in depth. I may check in a SQL file that only performs a read of the current schema structure, just for testability. The app-interface MRs might be follow on items.

chambridge commented:
@thearifismail I've referenced this information in the top level README. I created the README file in the `app_migrations` directory initially as a placeholder mechanism as you cannot check-in empty directories to git and I like the argument that a README in the directory is better than a `.gitkeep` file.

I don't know of any reason why this would need to wait to go in.

thearifismail reviewed (dismissed):
LGTM!

thearifismail reviewed (commented):
What is provided looks good to me but finding hard to picture without having a migration script, i.e. running a migration script

fstavela reviewed (approved):
LGTM! Thank you for the makefile changes and the additional documentation.

fstavela commented on README.md:
````diff
@@ -350,6 +350,18 @@ Run this command to generate a new revision in `migrations/versions`
+When migrations are being made it makes sense to capture the HBI schema state for replication subscribers to simplify their procedure for onboarding or schema recreation. Run the following command to capture the updated schema state:
+
+```bash
+make gen_hbi_schema_dump
````
We should mention that we need to install `postgresql16` to get the `pg_dump` command. We can add it at the [beginning of the README](https://github.com/RedHatInsights/insights-host-inventory?tab=readme-ov-file#pg_config) (the change is probably required only on Fedora and Debian docs)

chambridge replied:
I've got a Mac. What should the update look like if you wouldn't mind?

chambridge replied:
I pushed a couple changes.

fstavela replied:
When I installed it on a Fedora I used `sudo dnf install postgresql16` to make sure I have v16, but I just tried that installing `postgresql` also installs v16 on Fedora 40, so it's OK as well.

fstavela reviewed (commented):
I had a couple of comments, but otherwise, it looks good, thanks!

Just to make sure I understand the migration procedure correctly. First, we need to create a new DB migration in `migrations/versions` by `make migrate_db` as we would normally do. Then, we can use this migration and the revs in the `make gen_offline_sql` to generate an SQL migration script for the apps. Then (this is what I'm not sure about) we have to delete the original migration file from `migrations/versions` before committing because otherwise the migration would be executed on the publisher DB in Stage before apps get a chance to upgrade their subscriber DBs? And once all of the subscribers ran the migration, we can add the file back to `migrations/versions`.

chambridge commented on app_migrations/README.md:
````diff
@@ -0,0 +1,42 @@
+### Execution
+In the case below we are using the `query_hbi_schema.sql` migration file present in the `app_migrations` directory.
+```
+INVENTORY_MIGRATION_DRYRUN="true" INVENTORY_MIGRATION_FILE="query_hbi_schema.sql" python inv_migration_runner.py
````
yeah. Copy paste 😏

fstavela reviewed (dismissed):
Looks good now, thanks.

thearifismail commented on inv_migration_runner.py:
```diff
@@ -0,0 +1,78 @@
+    logger.exception("Inventory migration failed", exc_info=value)
+
+
+def run(config, logger, session, shutdown_handler, application):
```
Is the config parameter still needed, if not called in this function?

thearifismail commented on inv_migration_runner.py:
```diff
@@ -0,0 +1,78 @@
+        with open(file_path) as file:
+            sql_statements = file.read()
+            logger.info(f"Executing the following SQL statements:\n{sql_statements}")
+            if dry_run.lower() != "true":
```
Is a `dry_run` really needed if it does not execute any SQL statements?  Shouldn't a dry run execute SQL statements but NOT commit them or roll them back.

chambridge replied:
The SQL file selected could contain a COMMIT statement in it, so there would be no way of avoiding that. This is just an operational mechanism to perform the JOB and validate the SQL that would be run without running it. Allows you to prove out the deployment step and that the image of the container has the version of the SQL you intended to run.

thearifismail commented on app_migrations/hbi_schema_2024-11-15.sql:
```diff
@@ -0,0 +1,357 @@
+--
```
Should this file be included?

chambridge replied:
The intention was to supply the existing HBI schema, yes.

thearifismail commented on app/config.py:
```diff
@@ -78,7 +78,11 @@ def clowder_config(self):
 
         def topic(t):
-            return app_common_python.KafkaTopics[t].name if t else None
+            if t:
```
Is it related to this PR?

chambridge replied:
Yes. This ClowdJobInvocation fails if this isn't changed as no kafka topics are defined or needed for it.
